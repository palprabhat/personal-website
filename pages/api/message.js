const FormData = require("form-data");

const validateHuman = async (token) => {
  const secret = process.env.RECAPTCHA_SITE_SECRET;
  const response = await fetch(
    `https://www.google.com/recaptcha/api/siteverify?secret=${secret}&response=${token}`,
    {
      method: "POST",
    }
  );
  const data = await response.json();
  return data.success;
};

export default async function handler(req, res) {
  if (req.method === "POST") {
    const reqData = JSON.parse(req.body);

    const human = await validateHuman(reqData.reToken);
    console.log(human);
    if (!human) {
      res.status(400);
      res.json({ errors: ["Please, you're not fooling us, bot."] });
      return;
    }

    try {
      const formData = new FormData();

      Object.entries(reqData).map(([key, value]) => {
        formData.append(key, value);
      });

      const response = await fetch(process.env.CONTACT_API_URL, {
        method: "POST",
        body: formData,
      });
      const responseData = await response.json();

      if (responseData.result === "success") {
        res.status(200).send({ data: responseData.data });
        return;
      }

      res.status(500).send({ error: "Unable to send message" });
    } catch (err) {
      console.error(err);
      res.status(500).send({ error: "Unable to send message" });
    }
  } else {
    res.status(405).send({ error: "Request method not supported" });
  }
}
