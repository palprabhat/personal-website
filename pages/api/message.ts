/* eslint-disable @typescript-eslint/no-explicit-any */
import { NextApiRequest, NextApiResponse } from "next";
import FormData from "form-data";

type RequestBody = {
  name: string;
  email: string;
  subject: string;
  message: string;
  reToken: string;
};

type SendMessage = (reqBody: RequestBody) => Promise<string | null>;
type ValidateHuman = (token: string) => Promise<boolean>;

const cleanFromData = (formData: URLSearchParams) => {
  formData.forEach((_value, key) => formData.delete(key));
  formData.delete("_valueLength");
  formData.delete("writable");
  formData.delete("dataSize");
  formData.delete("pauseStreams");
  formData.delete("_streams");
  formData.delete("_insideLoop");
};

const validateHuman: ValidateHuman = async (token) => {
  const secret = process.env.RECAPTCHA_SITE_SECRET;
  const response = await fetch(
    `https://www.google.com/recaptcha/api/siteverify?secret=${secret}&response=${token}`,
    {
      method: "POST",
    }
  );
  const data = await response.json();
  return data.success as boolean;
};

const sendMessage: SendMessage = async (reqBody): Promise<any> => {
  try {
    const formData = new URLSearchParams(new FormData() as any);

    cleanFromData(formData);

    Object.entries(reqBody).map(([key, value]) => {
      if (key !== "reToken") formData.append(key, value);
    });

    const response = await fetch(process.env.CONTACT_API_URL, {
      method: "POST",
      body: formData,
    });

    const responseData = await response.json();

    if (responseData.result === "success") {
      return responseData.data;
    }
    return null;
  } catch (err) {
    console.error(err);
    return null;
  }
};

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
): Promise<void> {
  if (req.method === "POST") {
    const reqBody = JSON.parse(req.body) as RequestBody;

    const human = await validateHuman(reqBody.reToken);
    if (!human) {
      res.status(400).send({ error: "Could not verify you as human" });
      return;
    }

    const resData = await sendMessage(reqBody);

    if (resData) {
      res.status(200).send({ data: resData });
    } else {
      res.status(500).send({ error: "Unable to send message" });
    }
  } else {
    res.status(405).send({ error: "Request method not supported" });
  }
}
