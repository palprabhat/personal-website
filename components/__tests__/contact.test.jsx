import { render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Contact from "../contact";
import { mockFetch, mockMessage } from "../../pages/api/__mock__/message.mock";

const renderContactFormComponent = () => {
  const { getByText, getByPlaceholderText, ...utils } = render(<Contact />);

  const name = getByPlaceholderText("Name");
  const email = getByPlaceholderText("Email");
  const subject = getByPlaceholderText("Subject");
  const message = getByPlaceholderText("Message");
  const submitButton = getByText("Send message");

  const fillForm = () => {
    userEvent.type(name, mockMessage.name);
    userEvent.type(email, mockMessage.email);
    userEvent.type(subject, mockMessage.subject);
    userEvent.type(message, mockMessage.message);
  };

  return {
    name,
    email,
    subject,
    message,
    submitButton,
    fillForm,
    ...utils,
  };
};

beforeAll(() => jest.spyOn(window, "fetch"));
beforeEach(() => window.fetch.mockImplementation(mockFetch));
afterEach(() => window.fetch.mockClear());

describe("Contact Component", () => {
  it("matched snapshot", async () => {
    const { baseElement } = renderContactFormComponent();

    expect(baseElement).toMatchSnapshot();
  });

  it("should show thank you message on successful for submission", async () => {
    const {
      submitButton,
      fillForm,
      queryByText,
      baseElement,
    } = renderContactFormComponent();

    expect(submitButton).toBeInTheDocument();

    fillForm();
    userEvent.click(submitButton);

    await waitFor(() => {
      const thankyouMessage = queryByText(
        "Thank you for reaching out to me, I will get back to you soon."
      );
      expect(thankyouMessage).toBeInTheDocument();
    });

    expect(baseElement).toMatchSnapshot();
  });

  it("should show contact from on click on 'Send another message'", async () => {
    const {
      submitButton,
      fillForm,
      queryByText,
      baseElement,
    } = renderContactFormComponent();

    expect(submitButton).toBeInTheDocument();

    fillForm();
    userEvent.click(submitButton);

    await waitFor(async () => {
      const sendAnotherMessage = queryByText("Send another message");
      expect(sendAnotherMessage).toBeInTheDocument();

      userEvent.click(sendAnotherMessage);

      await waitFor(() => {
        const submitButton = queryByText("Send message");
        expect(submitButton).toBeInTheDocument();
      });

      expect(baseElement).toMatchSnapshot();
    });
  });
});
