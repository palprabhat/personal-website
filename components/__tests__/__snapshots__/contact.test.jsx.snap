// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact Component matched snapshot 1`] = `
<div>
  <section
    class="text-center section-p bg-primary-100"
  >
    <h2>
      Have a question or want to work together?
    </h2>
    <div
      class="max-w-6xl mx-auto mt-16 flex justify-between items-center"
    >
      <div
        class="w-1/2 hidden md:block"
      >
        <div
          style="max-width: 100%; width: 936px;"
        >
          <div
            style="position: relative; padding-bottom: 52.45726495726496%;"
          >
            
            <img
              alt="Contact"
              class="__lazy"
              data-src="/_next/image?url=%2Fimages%2Fcontact.svg&w=1024&q=75"
              data-srcset="/_next/image?url=%2Fimages%2Fcontact.svg&w=320&q=75 320w, /_next/image?url=%2Fimages%2Fcontact.svg&w=420&q=75 420w, /_next/image?url=%2Fimages%2Fcontact.svg&w=768&q=75 768w, /_next/image?url=%2Fimages%2Fcontact.svg&w=1024&q=75 1024w"
              style="visibility: hidden; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
            />
          </div>
        </div>
      </div>
      <div
        class="w-full md:w-1/2 md:px-8 flex flex-col justify-end"
        style="height: 450px;"
      >
        <form
          class="w-full undefined"
        >
          <div
            class="my-2 undefined"
          >
            <input
              aria-label="Name"
              class="border-none rounded-3xl w-full px-5 py-2 outline-none undefined"
              name="name"
              placeholder="Name"
            />
            <div
              class="text-red-600 text-sm text-left pl-5 h-6"
            />
          </div>
          <div
            class="my-2 undefined"
          >
            <input
              aria-label="Email"
              class="border-none rounded-3xl w-full px-5 py-2 outline-none undefined"
              name="email"
              placeholder="Email"
            />
            <div
              class="text-red-600 text-sm text-left pl-5 h-6"
            />
          </div>
          <div
            class="my-2 undefined"
          >
            <input
              aria-label="Subject"
              class="border-none rounded-3xl w-full px-5 py-2 outline-none undefined"
              name="subject"
              placeholder="Subject"
            />
            <div
              class="text-red-600 text-sm text-left pl-5 h-6"
            />
          </div>
          <div
            class="my-2 undefined"
          >
            <textarea
              aria-label="Message"
              class="border-none rounded-3xl px-4 py-2 w-full outline-none resize-none undefined"
              name="message"
              placeholder="Message"
              rows="3"
            />
            <div
              class="text-red-600 text-sm text-left pl-5 h-6"
            />
          </div>
          <div />
          <button
            class="btn-primary"
            type="submit"
          >
            Send message
          </button>
        </form>
        <div
          class="flex flex-wrap justify-around mt-8"
        >
          <a
            class="flex justify-center items-center px-2 group"
            data-testid="my-email"
            href="mailto:prabhatpal.14@gmail.com"
          >
            <span
              class="brand-icon-email mr-1 group-hover:text-primary-600 group-focus:text-primary-600"
            />
            prabhatpal.14@gmail.com
          </a>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`Contact Component should show contact from on click on 'Send another message' 1`] = `
<div>
  <section
    class="text-center section-p bg-primary-100"
  >
    <h2>
      Have a question or want to work together?
    </h2>
    <div
      class="max-w-6xl mx-auto mt-16 flex justify-between items-center"
    >
      <div
        class="w-1/2 hidden md:block"
      >
        <div
          style="max-width: 100%; width: 936px;"
        >
          <div
            style="position: relative; padding-bottom: 52.45726495726496%;"
          >
            
            <img
              alt="Contact"
              class="__lazy"
              data-src="/_next/image?url=%2Fimages%2Fcontact.svg&w=1024&q=75"
              data-srcset="/_next/image?url=%2Fimages%2Fcontact.svg&w=320&q=75 320w, /_next/image?url=%2Fimages%2Fcontact.svg&w=420&q=75 420w, /_next/image?url=%2Fimages%2Fcontact.svg&w=768&q=75 768w, /_next/image?url=%2Fimages%2Fcontact.svg&w=1024&q=75 1024w"
              style="visibility: hidden; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
            />
          </div>
        </div>
      </div>
      <div
        class="w-full md:w-1/2 md:px-8 flex flex-col justify-end"
        style="height: 450px;"
      >
        <form
          class="w-full undefined"
        >
          <div
            class="my-2 undefined"
          >
            <input
              aria-label="Name"
              class="border-none rounded-3xl w-full px-5 py-2 outline-none undefined"
              name="name"
              placeholder="Name"
            />
            <div
              class="text-red-600 text-sm text-left pl-5 h-6"
            />
          </div>
          <div
            class="my-2 undefined"
          >
            <input
              aria-label="Email"
              class="border-none rounded-3xl w-full px-5 py-2 outline-none undefined"
              name="email"
              placeholder="Email"
            />
            <div
              class="text-red-600 text-sm text-left pl-5 h-6"
            />
          </div>
          <div
            class="my-2 undefined"
          >
            <input
              aria-label="Subject"
              class="border-none rounded-3xl w-full px-5 py-2 outline-none undefined"
              name="subject"
              placeholder="Subject"
            />
            <div
              class="text-red-600 text-sm text-left pl-5 h-6"
            />
          </div>
          <div
            class="my-2 undefined"
          >
            <textarea
              aria-label="Message"
              class="border-none rounded-3xl px-4 py-2 w-full outline-none resize-none undefined"
              name="message"
              placeholder="Message"
              rows="3"
            />
            <div
              class="text-red-600 text-sm text-left pl-5 h-6"
            />
          </div>
          <div />
          <button
            class="btn-primary"
            type="submit"
          >
            Send message
          </button>
        </form>
        <div
          class="flex flex-wrap justify-around mt-8"
        >
          <a
            class="flex justify-center items-center px-2 group"
            data-testid="my-email"
            href="mailto:prabhatpal.14@gmail.com"
          >
            <span
              class="brand-icon-email mr-1 group-hover:text-primary-600 group-focus:text-primary-600"
            />
            prabhatpal.14@gmail.com
          </a>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`Contact Component should show thank you message on successful for submission 1`] = `
<div>
  <section
    class="text-center section-p bg-primary-100"
  >
    <h2>
      Have a question or want to work together?
    </h2>
    <div
      class="max-w-6xl mx-auto mt-16 flex justify-between items-center"
    >
      <div
        class="w-1/2 hidden md:block"
      >
        <div
          style="max-width: 100%; width: 936px;"
        >
          <div
            style="position: relative; padding-bottom: 52.45726495726496%;"
          >
            
            <img
              alt="Contact"
              class="__lazy"
              data-src="/_next/image?url=%2Fimages%2Fcontact.svg&w=1024&q=75"
              data-srcset="/_next/image?url=%2Fimages%2Fcontact.svg&w=320&q=75 320w, /_next/image?url=%2Fimages%2Fcontact.svg&w=420&q=75 420w, /_next/image?url=%2Fimages%2Fcontact.svg&w=768&q=75 768w, /_next/image?url=%2Fimages%2Fcontact.svg&w=1024&q=75 1024w"
              style="visibility: hidden; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
            />
          </div>
        </div>
      </div>
      <div
        class="w-full md:w-1/2 md:px-8 flex flex-col justify-end"
        style="height: 450px;"
      >
        <div>
          <h3
            class="mb-24"
          >
            Thank you for reaching out to me, I will get back to you soon.
          </h3>
          <button
            class="btn-secondary"
            type="button"
          >
            Send another message
          </button>
        </div>
        <div
          class="flex flex-wrap justify-around mt-8"
        >
          <a
            class="flex justify-center items-center px-2 group"
            data-testid="my-email"
            href="mailto:prabhatpal.14@gmail.com"
          >
            <span
              class="brand-icon-email mr-1 group-hover:text-primary-600 group-focus:text-primary-600"
            />
            prabhatpal.14@gmail.com
          </a>
        </div>
      </div>
    </div>
  </section>
</div>
`;
